File: android/ontfeatures/src/main/kotlin/com/labibleont/ont/features/reading/ReferencePicker.kt
package com.labibleont.ont.features.reading

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.MenuBook
import androidx.compose.material.icons.automirrored.filled.Notes
import androidx.compose.material3.Icon
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.labibleont.ont.designsystem.metrics.ONTRadius
import com.labibleont.ont.designsystem.metrics.ontSpacing
import com.labibleont.ont.designsystem.surfaces.ONTGroup
import com.labibleont.ont.designsystem.surfaces.ONTGroupDivider
import com.labibleont.ont.designsystem.surfaces.ONTPage
import com.labibleont.ont.designsystem.surfaces.SectionCaption
import com.labibleont.ont.designsystem.theme.LocalReadingTheme
import com.labibleont.ont.designsystem.tokens.ONTColors
import com.labibleont.ont.designsystem.typography.ONTFonts
import com.labibleont.ont.kit.corpus.BookOutline
import com.labibleont.ont.kit.corpus.ChapterStub
import com.labibleont.ont.kit.corpus.Status
import com.labibleont.ont.kit.reader.ReadingFont
import com.labibleont.ont.kit.search.SearchEngine

/**
 * Le sélecteur de renvoi — livre, unité, verset.
 *
 * ## Ce qu'il remplace
 *
 * Le geste de YouVersion et de Bible Strong : une pastille qui dit où l'on est,
 * et qu'on touche pour aller ailleurs. Sans lui, aller de *Bereshit* 1 à
 * *Bereshit* 18 demandait quatre gestes — revenir au sommaire, replier le
 * livre, le déplier, viser. C'est exactement le manque signalé : toucher un
 * livre ouvrait la première unité et il n'y avait pas d'autre chemin.
 *
 * ## Trois étapes, dont on sort à la deuxième
 *
 * Livres, unités, versets. Neuf fois sur dix on s'arrête à l'unité, alors
 * chaque étape porte sa sortie courte en tête — « Introduction », « Toute
 * l'unité » — plutôt que de forcer à traverser la troisième.
 *
 * ## Il lit l'esquisse, pas le livre
 *
 * `books/bereshit.json` fait 750 Ko d'arbre d'inline. L'esquisse porte déjà le
 * numéro, le titre, le statut et le nombre de versets de chaque unité : tout ce
 * qu'une grille de numéros demande. Ouvrir le livre pour la dessiner serait
 * payer très cher un renseignement qu'on a.
 */
@Composable
public fun ReferencePicker(
    model: ReadingModel,
    /** L'unité ouverte — le sélecteur s'ouvre là, pas en haut de la liste. */
    livreCourant: String?,
    uniteCourante: String?,
    /**
     * Un livre imposé à l'ouverture, distinct de celui qu'on lit.
     *
     * C'est par là que passe l'onglet Bible : y toucher un livre doit montrer
     * **ses** unités, pas celles du chapitre ouvert ailleurs — et surtout pas
     * sauter directement à la première, ce qu'il faisait.
     */
    livreImpose: String? = null,
    onAller: (bookId: String, chapterId: String, verse: Int?) -> Unit,
    modifier: Modifier = Modifier,
) {
    // Un chemin, pas des onglets : on avance et on revient. La pile est tenue
    // ici plutôt que par la navigation de l'app, parce qu'elle ne sort pas de
    // ce sélecteur — l'app n'a pas à connaître ses étapes.
    var etape: Etape by remember(livreCourant, livreImpose) {
        mutableStateOf(
            when {
                // Un livre imposé l'emporte : il vient d'un geste explicite.
                livreImpose != null -> Etape.Unites(livreImpose)
                // Sinon on ouvre sur le livre courant : le lecteur cherche
                // presque toujours à côté de là où il est.
                livreCourant != null -> Etape.Unites(livreCourant)
                else -> Etape.Livres
            },
        )
    }

    when (val e = etape) {
        Etape.Livres -> EtapeDesLivres(
            model = model,
            livreCourant = livreCourant,
            onChoisir = { etape = Etape.Unites(it) },
            modifier = modifier,
        )

        is Etape.Unites -> EtapeDesUnites(
            livre = model.esquisse(e.livre),
            uniteCourante = uniteCourante,
            onRetour = { etape = Etape.Livres },
            onUnite = { unite -> etape = Etape.Versets(e.livre, unite) },
            onIntroduction = { intro -> onAller(e.livre, intro, null) },
            modifier = modifier,
        )

        is Etape.Versets -> EtapeDesVersets(
            unite = model.esquisse(e.livre)?.chapters?.firstOrNull { it.id == e.unite },
            francaisRecu = model.preferences.french,
            onRetour = { etape = Etape.Unites(e.livre) },
            onToutLUnite = { onAller(e.livre, e.unite, null) },
            onVerset = { n -> onAller(e.livre, e.unite, n) },
            modifier = modifier,
        )
    }
}

private sealed interface Etape {
    data object Livres : Etape
    data class Unites(val livre: String) : Etape
    data class Versets(val livre: String, val unite: String) : Etape
}

// ── 1. Les livres ───────────────────────────────────────────────────────

@Composable
private fun EtapeDesLivres(
    model: ReadingModel,
    livreCourant: String?,
    onChoisir: (String) -> Unit,
    modifier: Modifier,
) {
    val theme = LocalReadingTheme.current
    val espace = ontSpacing
    var recherche by remember { mutableStateOf("") }

    Column(modifier = modifier.fillMaxWidth().verticalScroll(rememberScrollState())) {
        ONTPage {
            OutlinedTextField(
                value = recherche,
                onValueChange = { recherche = it },
                placeholder = { Text("Chercher un livre") },
                singleLine = true,
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
                colors = TextFieldDefaults.colors(
                    focusedContainerColor = ONTColors.surface(theme),
                    unfocusedContainerColor = ONTColors.surface(theme),
                    focusedIndicatorColor = ONTColors.accent(theme),
                    unfocusedIndicatorColor = ONTColors.separator(theme),
                    cursorColor = ONTColors.accent(theme),
                ),
                modifier = Modifier.fillMaxWidth().padding(vertical = espace.s),
            )

            for (corpus in model.corpora.sortedBy { it.order }) {
                val livres = corpus.modes.sortedBy { it.order }
                    .flatMap { it.books }
                    .filter { correspond(it, recherche) }
                if (livres.isEmpty()) continue

                Spacer(Modifier.height(espace.m))
                SectionCaption(corpus.title, teinte = ONTColors.brandInk(theme))
                ONTGroup {
                    livres.forEachIndexed { i, livre ->
                        if (i > 0) ONTGroupDivider()
                        LigneDuSelecteur(
                            livre = livre,
                            courant = livre.id == livreCourant,
                            onChoisir = { onChoisir(livre.id) },
                        )
                    }
                }
            }
            Spacer(Modifier.height(espace.xxl))
        }
    }
}

/**
 * Replie une chaîne pour la comparaison : sans accents, sans casse.
 *
 * On emploie le `fold` du moteur de recherche plutôt qu'un `lowercase` : c'est
 * le même pliage que celui du pipeline, donc chercher « berechit » trouve
 * *Bereshit* et « genese » trouve *Genèse* — ce qu'un lecteur tape vraiment.
 */
private fun correspond(livre: BookOutline, recherche: String): Boolean {
    val cherche = SearchEngine.fold(recherche.trim())
    if (cherche.isEmpty()) return true
    return SearchEngine.fold(livre.title).contains(cherche) ||
        SearchEngine.fold(livre.french).contains(cherche)
}

@Composable
private fun LigneDuSelecteur(livre: BookOutline, courant: Boolean, onChoisir: () -> Unit) {
    val theme = LocalReadingTheme.current
    val espace = ontSpacing
    // Un slot vide reste **visible mais éteint** : le corpus est un chantier, et
    // masquer les vides donnerait une fausse idée de sa forme.
    val lisible = !livre.empty

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .then(if (lisible) Modifier.clickable(onClick = onChoisir) else Modifier)
            .heightIn(min = 48.dp)
            .padding(horizontal = espace.l, vertical = espace.s),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            "%02d".format(livre.slot),
            fontSize = 11.sp,
            color = ONTColors.inkSoft(theme),
            modifier = Modifier.padding(end = espace.m),
        )
        Column(modifier = Modifier.weight(1f)) {
            Text(
                livre.title,
                fontFamily = ONTFonts.family(ReadingFont.LITERATA),
                fontStyle = FontStyle.Italic,
                fontSize = 17.sp,
                fontWeight = if (courant) FontWeight.SemiBold else FontWeight.Normal,
                color = when {
                    courant -> ONTColors.accent(theme)
                    lisible -> ONTColors.ink(theme)
                    else -> ONTColors.inkSoft(theme)
                },
            )
            Text(livre.french, fontSize = 13.sp, color = ONTColors.inkSoft(theme))
        }
        if (!lisible) {
            Text("à venir", fontSize = 12.sp, color = ONTColors.inkSoft(theme))
        }
    }
}

// ── 2. Les unités ───────────────────────────────────────────────────────

@Composable
private fun EtapeDesUnites(
    livre: BookOutline?,
    uniteCourante: String?,
    onRetour: () -> Unit,
    onUnite: (String) -> Unit,
    onIntroduction: (String) -> Unit,
    modifier: Modifier,
) {
    val theme = LocalReadingTheme.current
    val espace = ontSpacing

    Column(modifier = modifier.fillMaxWidth()) {
        FilDAriane(livre?.title ?: "", onRetour = onRetour)

        if (livre == null || livre.empty) {
            Indisponible("Ce livre n'est pas encore rédigé.")
            return@Column
        }

        livre.intro?.let { intro ->
            SortieCourte(
                intitule = "Introduction",
                icone = Icons.AutoMirrored.Filled.MenuBook,
                onClick = { onIntroduction(intro.id) },
                modifier = Modifier.padding(horizontal = espace.l),
            )
            Spacer(Modifier.height(espace.m))
        }

        LazyVerticalGrid(
            columns = GridCells.Adaptive(minSize = 64.dp),
            horizontalArrangement = Arrangement.spacedBy(espace.s),
            verticalArrangement = Arrangement.spacedBy(espace.s),
            contentPadding = androidx.compose.foundation.layout.PaddingValues(espace.l),
        ) {
            items(livre.chapters, key = { it.id }) { unite ->
                Case(
                    unite = unite,
                    courant = unite.id == uniteCourante,
                    onClick = { onUnite(unite.id) },
                )
            }
        }
    }
}

/**
 * Une case de la grille.
 *
 * Le numéro, et rien d'autre — le titre d'une unité ONT est une phrase, elle ne
 * tient pas dans une case. Un brouillon porte un point : c'est l'état du §12,
 * et le lecteur doit savoir avant d'ouvrir que ce texte n'est pas encore
 * verrouillé.
 */
@Composable
private fun Case(unite: ChapterStub, courant: Boolean, onClick: () -> Unit) {
    val theme = LocalReadingTheme.current

    Box(
        modifier = Modifier
            .size(64.dp)
            .clip(RoundedCornerShape(ONTRadius.highlight * 2))
            .background(
                if (courant) ONTColors.accent(theme).copy(alpha = 0.25f)
                else ONTColors.surface(theme),
            )
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            "${unite.n}",
            fontFamily = ONTFonts.display,
            fontSize = 19.sp,
            color = if (courant) ONTColors.accent(theme) else ONTColors.ink(theme),
            textAlign = TextAlign.Center,
        )
        if (unite.status == Status.BROUILLON) {
            Box(
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .padding(6.dp)
                    .size(6.dp)
                    .clip(RoundedCornerShape(ONTRadius.pill))
                    .background(ONTColors.accentuation(theme)),
            )
        }
    }
}

// ── 3. Les versets ──────────────────────────────────────────────────────

@Composable
private fun EtapeDesVersets(
    unite: ChapterStub?,
    francaisRecu: Boolean,
    onRetour: () -> Unit,
    onToutLUnite: () -> Unit,
    onVerset: (Int) -> Unit,
    modifier: Modifier,
) {
    val espace = ontSpacing

    Column(modifier = modifier.fillMaxWidth()) {
        // **Le même mot que le sommaire.** Le fil d'Ariane disait le titre brut
        // — « Bereshit 2 » — pendant que l'arborescence disait déjà
        // « Parashah 2 ». Deux noms pour la même unité, à un écran d'écart.
        FilDAriane(unite?.libelle(francaisRecu) ?: "", onRetour = onRetour)

        // La sortie courte, en premier : neuf fois sur dix on veut l'unité, pas
        // un verset précis. Elle nomme l'unité dans le registre choisi plutôt
        // que de la dire « unité » — le mot générique était le seul endroit de
        // l'app où le lecteur ne lisait ni « chapitre » ni « parashah ».
        SortieCourte(
            intitule = if (francaisRecu) "Tout le chapitre" else "Toute la parashah",
            icone = Icons.AutoMirrored.Filled.Notes,
            onClick = onToutLUnite,
            modifier = Modifier.padding(horizontal = espace.l),
        )

        if (unite == null || unite.verseCount == 0) {
            Indisponible("Cette unité ne porte pas de versets.")
            return@Column
        }

        Spacer(Modifier.height(espace.m))
        LazyVerticalGrid(
            columns = GridCells.Adaptive(minSize = 56.dp),
            horizontalArrangement = Arrangement.spacedBy(espace.s),
            verticalArrangement = Arrangement.spacedBy(espace.s),
            contentPadding = androidx.compose.foundation.layout.PaddingValues(espace.l),
        ) {
            items((1..unite.verseCount).toList()) { n ->
                CaseDeVerset(n = n, onClick = { onVerset(n) })
            }
        }
    }
}

@Composable
private fun CaseDeVerset(n: Int, onClick: () -> Unit) {
    val theme = LocalReadingTheme.current
    Box(
        modifier = Modifier
            .size(56.dp)
            .clip(RoundedCornerShape(ONTRadius.highlight * 2))
            .background(ONTColors.surface(theme))
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Text("$n", fontSize = 16.sp, color = ONTColors.ink(theme))
    }
}

// ── Le commun ───────────────────────────────────────────────────────────

/**
 * Le fil d'Ariane d'une étape.
 *
 * Il porte le retour **dans** le sélecteur, distinct du retour système qui,
 * lui, ferme le sélecteur entier. Les deux existent parce qu'ils ne font pas la
 * même chose : revenir d'une étape n'est pas renoncer à changer de passage.
 */
@Composable
private fun FilDAriane(titre: String, onRetour: () -> Unit) {
    val theme = LocalReadingTheme.current
    val espace = ontSpacing
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onRetour)
            .padding(horizontal = espace.l, vertical = espace.m),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            Icons.AutoMirrored.Filled.ArrowBack,
            contentDescription = "Revenir à l'étape précédente",
            tint = ONTColors.brandInk(theme),
            modifier = Modifier.padding(end = espace.s),
        )
        Text(
            titre,
            fontFamily = ONTFonts.display,
            fontSize = 18.sp,
            color = ONTColors.inkStrong(theme),
        )
    }
}

@Composable
private fun SortieCourte(
    intitule: String,
    icone: androidx.compose.ui.graphics.vector.ImageVector,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val theme = LocalReadingTheme.current
    val espace = ontSpacing
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(ONTRadius.card))
            .background(ONTColors.accent(theme).copy(alpha = 0.10f))
            .clickable(onClick = onClick)
            .padding(espace.m),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            icone,
            contentDescription = null,
            tint = ONTColors.accent(theme),
            modifier = Modifier.padding(end = espace.s),
        )
        Text(intitule, color = ONTColors.accent(theme), fontSize = 15.sp)
    }
}

@Composable
private fun Indisponible(message: String) {
    val theme = LocalReadingTheme.current
    Text(
        message,
        color = ONTColors.inkSoft(theme),
        modifier = Modifier.fillMaxWidth().padding(ontSpacing.xl),
        textAlign = TextAlign.Center,
    )
}

File: app/Packages/ONTFeatures/Sources/ReadingFeature/Presentation/ReferencePicker.swift
import ONTDesignSystem
import ONTKit
import SwiftUI

/// Le sélecteur de renvoi — livre, unité, verset.
///
/// Le geste de YouVersion et de Bible Strong : une pastille en haut à gauche
/// qui dit où l'on est, et qu'on touche pour aller ailleurs. Trois étapes,
/// parce qu'il y a trois choses à choisir — mais on peut s'arrêter à la
/// deuxième, et c'est le cas courant.
///
/// Ce qu'il remplace : revenir à la table des matières, replier un livre,
/// en déplier un autre. Quatre gestes pour aller de *Bereshit* 1 à
/// *Bereshit* 18.
public struct ReferencePicker: View {
    @Environment(ReadingModel.self) private var model
    @Environment(Router.self) private var router
    @Environment(\.dismiss) private var dismiss
    @Environment(\.ontTheme) private var theme

    var spacing = ONTSpacing()
    var echelle = ONTScaled()

    /// L'unité ouverte — le sélecteur s'ouvre là, pas en haut de la liste.
    ///
    /// **Facultative**, parce qu'on entre ici par deux portes. Depuis la
    /// lecture, il y a une unité courante et le sélecteur la marque. Depuis le
    /// sommaire d'un livre, il n'y en a pas : on n'a rien ouvert, on choisit.
    private let current: Chapter?

    /// L'étape en cours. Un chemin, pas des onglets : on avance et on revient.
    private enum Etape: Hashable {
        case unites(book: String)
        case versets(book: String, chapter: String)
    }

    /// Où la feuille s'ouvre.
    private let depart: Etape?

    public init(current: Chapter) {
        self.current = current
        self.depart = nil
    }

    /// Le sélecteur ouvert **directement sur les versets** d'une unité.
    ///
    /// C'est la porte du sommaire : le lecteur y a déjà choisi son livre et son
    /// unité, la seule chose qui lui reste à dire est *où commencer*. Le faire
    /// repasser par les deux étapes précédentes serait lui redemander ce qu'il
    /// vient de répondre.
    public init(book: String, chapter: String) {
        self.current = nil
        self.depart = .versets(book: book, chapter: chapter)
    }

    @State private var chemin: [Etape] = []
    @State private var recherche = ""

    public var body: some View {
        NavigationStack(path: $chemin) {
            livres
                .navigationTitle("Aller à")
                .navigationBarTitleDisplayMode(.inline)
                .navigationDestination(for: Etape.self) { etape in
                    switch etape {
                    case .unites(let book):
                        unites(of: book)
                    case .versets(let book, let chapter):
                        versets(book: book, chapter: chapter)
                    }
                }
                .toolbar {
                    ToolbarItem(placement: .cancellationAction) {
                        Button("Fermer") { dismiss() }
                    }
                }
        }
        .onAppear {
            // On ouvre sur le livre courant : le lecteur cherche presque
            // toujours à côté de là où il est.
            chemin = depart.map { [$0] } ?? [.unites(book: current?.bookId ?? "")]
        }
    }

    // MARK: - 1. Les livres

    private var livres: some View {
        List {
            ForEach(model.corpora) { corpus in
                let livresDuCorpus = corpus.modes
                    .sorted { $0.order < $1.order }
                    .flatMap(\.books)
                    .filter { correspond($0) }

                if !livresDuCorpus.isEmpty {
                    Section {
                        ForEach(livresDuCorpus) { livre in
                            Button {
                                chemin = [.unites(book: livre.id)]
                            } label: {
                                LivreLigne(livre: livre, courant: livre.id == current?.bookId)
                                    // Même défaut que « Reprendre » : le vide à
                                    // droite du titre ne répondait pas.
                                    .frame(maxWidth: .infinity, alignment: .leading)
                                    .contentShape(.rect)
                            }
                            .buttonStyle(.plain)
                            // Un slot vide reste **visible mais éteint** : le
                            // corpus est un chantier, et masquer les vides
                            // donnerait une fausse idée de sa forme.
                            .disabled(livre.empty)
                            .ontRow()
                        }
                    } header: {
                        Text(corpus.title)
                            .font(.custom(ONTFonts.display, size: 15))
                            .textCase(nil)
                            .foregroundStyle(ONTColors.brandInk(theme.mode))
                    }
                }
            }
        }
        .listStyle(.insetGrouped)
        .searchable(text: $recherche, prompt: "Chercher un livre")
        .ontScreen()
    }

    private func correspond(_ livre: BookOutline) -> Bool {
        guard !recherche.isEmpty else { return true }
        let cherche = replie(recherche)
        return replie(livre.title).contains(cherche)
            || replie(livre.french).contains(cherche)
    }

    /// Replie une chaîne pour la comparaison : sans accents, sans casse.
    ///
    /// Sans ça, chercher « berechit » ne trouve pas « Bereshit », et chercher
    /// « genese » ne trouve pas « Genèse » — ce qui est précisément ce qu'un
    /// lecteur tape.
    private func replie(_ texte: String) -> String {
        texte.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: .current)
    }

    // MARK: - 2. Les unités

    @ViewBuilder
    private func unites(of bookId: String) -> some View {
        // L'**esquisse** du livre, pas le livre : elle porte déjà le numéro, le
        // titre, le statut et le nombre de versets de chaque unité. Charger
        // `books/bereshit.json` — 750 Ko d'arbre d'inline — pour dessiner une
        // grille de numéros serait payer très cher un renseignement qu'on a.
        let livre = model.outline(bookId)
        ScrollView {
            if let livre, !livre.empty {
                VStack(alignment: .leading, spacing: spacing.m) {
                    if let intro = livre.intro {
                        Button {
                            aller(book: bookId, chapter: intro.id, verse: nil)
                        } label: {
                            Label("Introduction", systemImage: "text.book.closed")
                                .font(.callout.weight(.medium))
                                .foregroundStyle(theme.accent)
                                .frame(maxWidth: .infinity, alignment: .leading)
                                .padding(spacing.m)
                                .background(
                                    RoundedRectangle(cornerRadius: ONTRadius.card)
                                        .fill(theme.accent.opacity(0.10))
                                )
                                .contentShape(.rect)
                        }
                        .buttonStyle(.plain)
                    }

                    LazyVGrid(columns: grille, spacing: spacing.s) {
                        ForEach(livre.chapters) { unite in
                            Case(
                                titre: "\(unite.n)",
                                courant: unite.id == current?.id,
                                brouillon: unite.status == .brouillon
                            ) {
                                chemin.append(.versets(book: bookId, chapter: unite.id))
                            }
                        }
                    }
                }
                .padding(spacing.m)
            } else {
                Indisponible(message: "Ce livre n'est pas encore rédigé.")
            }
        }
        .navigationTitle(livre?.title ?? bookId)
        .navigationBarTitleDisplayMode(.inline)
        .ontScreen()
    }

    // MARK: - 3. Les versets

    @ViewBuilder
    private func versets(book: String, chapter: String) -> some View {
        let unite = model.outline(book)?.chapters.first { $0.id == chapter }
        ScrollView {
            VStack(alignment: .leading, spacing: spacing.m) {
                // La sortie courte, en premier : neuf fois sur dix on veut
                // l'unité, pas un verset précis.
                Button {
                    aller(book: book, chapter: chapter, verse: nil)
                } label: {
                    Label(toutLUnite, systemImage: "text.justify.left")
                        .font(.callout.weight(.medium))
                        .foregroundStyle(theme.accent)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .padding(spacing.m)
                        .background(
                            RoundedRectangle(cornerRadius: ONTRadius.card)
                                .fill(theme.accent.opacity(0.10))
                        )
                        .contentShape(.rect)
                }
                .buttonStyle(.plain)

                if let unite, unite.verseCount > 0 {
                    // **Le verset où l'on est se marque, comme le livre et
                    // l'unité aux deux étapes précédentes.**
                    //
                    // Il était câblé à `false` : les deux premières étapes
                    // disaient où l'on est, la troisième ne le disait pas, et
                    // le lecteur arrivait sur vingt et une cases identiques
                    // sans savoir laquelle il venait de quitter. VoiceOver ne
                    // l'annonçait pas non plus — `courant` porte aussi le trait
                    // `.isSelected`.
                    let courant = versetCourant(book: book, chapter: chapter)
                    LazyVGrid(columns: grille, spacing: spacing.s) {
                        ForEach(1...unite.verseCount, id: \.self) { n in
                            Case(titre: "\(n)", courant: n == courant, brouillon: false) {
                                aller(book: book, chapter: chapter, verse: n)
                            }
                        }
                    }
                }
            }
            .padding(spacing.m)
        }
        .navigationTitle(unite?.label(french: model.preferences.french) ?? chapter)
        .navigationBarTitleDisplayMode(.inline)
        .ontScreen()
    }

    /// Le verset que le lecteur quitte, s'il est dans l'unité qu'on regarde.
    ///
    /// La position mémorisée est **la sienne**, pas celle de l'unité affichée :
    /// on ne la retient que si les deux coïncident. Sans ce garde-fou, ouvrir
    /// une autre unité que celle qu'on lit allumerait une case au hasard —
    /// « verset 12 » de *Bereshit* 2 parce qu'on en était au 12 de *Bereshit*
    /// 18.
    private func versetCourant(book: String, chapter: String) -> Int? {
        guard let position = model.position,
              position.bookId == book,
              position.chapterId == chapter
        else { return nil }
        return position.verse
    }

    /// « Tout le chapitre » ou « Toute la parashah », selon le registre.
    ///
    /// Le libellé disait « Toute l'unité » — le mot du pipeline, juste et
    /// interne. Le lecteur, lui, vient de toucher « Chapitre 2 » ou
    /// « Parashah 2 » ; lui répondre « unité » introduit un troisième nom pour
    /// la même chose, au moment précis où le réglage cherche à n'en enseigner
    /// qu'un.
    private var toutLUnite: String {
        LibelleDUnite.toutLe(french: model.preferences.french)
    }

    private func aller(book: String, chapter: String, verse: Int?) {
        router.open(book: book, chapter: chapter, verse: verse)
        dismiss()
    }

    /// Le minimum suit le curseur, comme la case qu'il accueille — sinon des
    /// chiffres devenus plus larges débordent d'une colonne restée à 54. La
    /// grille étant `.adaptive`, monter le minimum ne casse rien : elle pose
    /// simplement moins de colonnes.
    private var grille: [GridItem] {
        [GridItem(.adaptive(minimum: echelle(54)), spacing: spacing.s)]
    }
}

// MARK: - Pièces

/// Une case de la grille — une unité ou un verset.
///
/// 54 points de côté au minimum : en dessous, on rate la cible en tenant le
/// téléphone d'une main, qui est la posture de lecture.
private struct Case: View {
    @Environment(\.ontTheme) private var theme
    var echelle = ONTScaled()

    let titre: String
    let courant: Bool
    let brouillon: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(titre)
                .font(.system(size: echelle(17), weight: courant ? .semibold : .regular))
                .monospacedDigit()
                .foregroundStyle(courant ? theme.surface : theme.ink)
                .frame(minWidth: echelle(54), minHeight: echelle(48))
                .background(
                    RoundedRectangle(cornerRadius: ONTRadius.card)
                        .fill(courant ? theme.accent : theme.ink.opacity(0.06))
                )
                .overlay(alignment: .topTrailing) {
                    // Un brouillon se signale sans se refuser : §12 dit qu'il
                    // ne fait pas référence, pas qu'on ne peut pas le lire.
                    if brouillon {
                        Circle()
                            .fill(ONTColors.accent(theme.mode))
                            .frame(width: 6, height: 6)
                            .padding(6)
                    }
                }
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(courant ? [.isButton, .isSelected] : .isButton)
    }
}

private struct LivreLigne: View {
    @Environment(\.ontTheme) private var theme

    let livre: BookOutline
    let courant: Bool

    var body: some View {
        HStack(spacing: 10) {
            Text("\(livre.slot)")
                .font(.caption.monospacedDigit())
                .foregroundStyle(theme.ink.opacity(0.35))
                .frame(width: 24, alignment: .trailing)

            VStack(alignment: .leading, spacing: 1) {
                Text(livre.title)
                    .font(.custom(ONTFonts.display, size: 16))
                    .foregroundStyle(livre.empty ? theme.ink.opacity(0.35) : theme.ink)
                Text(livre.french)
                    .font(.caption)
                    .foregroundStyle(theme.ink.opacity(0.45))
            }
            Spacer(minLength: 8)
            if courant {
                Image(systemName: "checkmark")
                    .font(.footnote.weight(.semibold))
                    .foregroundStyle(theme.accent)
            }
        }
        .contentShape(.rect)
    }
}

private struct Indisponible: View {
    @Environment(\.ontTheme) private var theme
    let message: String

    var body: some View {
        Text(message)
            .font(.callout)
            .foregroundStyle(theme.ink.opacity(0.55))
            .frame(maxWidth: .infinity)
            .padding(.top, 48)
    }
}
